src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py

```diff
@@ -278,28 +285,37 @@ def validate(self, node, options=None):
                 f"supplied node argument should be a PSyIR Range, but "
                 f"found '{type(node).__name__}'.")
         # Am I within an array reference?
-        if not node.parent or not isinstance(node.parent, ArrayReference):
+        if not node.parent or not isinstance(node.parent, ArrayMixin):
             raise TransformationError(
                 f"Error in NemoArrayRange2LoopTrans transformation. The "
-                f"supplied node argument should be within an ArrayReference "
+                f"supplied node argument should be within an ArrayMixin "
                 f"node, but found '{type(node.parent).__name__}'.")
-        array_ref = node.parent
         # Is the array reference within an assignment?
-        if not array_ref.parent or not isinstance(array_ref.parent,
-                                                  Assignment):
+        assignment = node.ancestor(Assignment)
+        if not assignment:
             raise TransformationError(
                 f"Error in NemoArrayRange2LoopTrans transformation. The "
-                f"supplied node argument should be within an ArrayReference "
-                f"node that is within an Assignment node, but found "
-                f"'{type(array_ref.parent).__name__}'.")
-        assignment = array_ref.parent
+                f"supplied node argument should be within an Assignment node, "
+                f"but found a '{node}' that is not in an assignment.")
         # Is the array reference the lhs of the assignment?
-        if assignment.lhs is not array_ref:
+        if node not in assignment.lhs.walk(Node):
             raise TransformationError(
                 "Error in NemoArrayRange2LoopTrans transformation. The "
-                "supplied node argument should be within an ArrayReference "
+                "supplied node argument should be within an ArrayMixin "
                 "node that is within the left-hand-side of an Assignment "
                 "node, but it is on the right-hand-side.")
+
+        # We don't support nested range expressions
+        for range_expr in assignment.walk(Range):
```

Comment:
I agree it shouldn't be valid and as first step the frontend shouldn't allow them to be constructed. However I am unsure if we can easily prevent a purposely misconstructed tree since this may not fall in the parent-child validation check and the global_validation check happen after the transformation. So it may be good to keep the check here (even if it is to check for inconsistent trees).

Also, this also check for this nesting `mystruct%field5(indices(:)) = 0.0d0`.

Reply:
Ah yes, thanks for explaining.